pkg/tenant/topics.go

```diff
@@ -237,24 +254,100 @@ func (s *Server) TopicUpdate(c *gin.Context) {
+		return
+	}
+
 	// Verify the topic name exists and return a 400 response if it doesn't.
 	if topic.Name == "" {
 		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic name is required"))
 		return
 	}
 
-	// Get the specified topic from the database and return a 404 response if
-	// it cannot be retrieved.
+	// Topic state is optional but can only be set to readonly.
+	if topic.State != "" && topic.State != pb.TopicTombstone_READONLY.String() {
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("topic state can only be set to READONLY"))
+		return
+	}
+
+	// Fetch the topic metadata from the database.
 	var t *db.Topic
-	if t, err = db.RetrieveTopic(c.Request.Context(), topicID); err != nil {
+	if t, err = db.RetrieveTopic(ctx, topicID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			log.Warn().Err(err).Str("topicID", topicID.String()).Msg("topic not found")
+			c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
+			return
+		}
 		log.Error().Err(err).Str("topicID", topicID.String()).Msg("could not retrieve topic")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not update topic"))
+		return
+	}
+
+	// Ensure the new name is valid
+	t.Name = topic.Name
+	if err = t.Validate(); err != nil {
+		log.Warn().Err(err).Msg("could not validate topic update")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+		return
+	}
+
+	// Verify that the user owns the topic
+	if claims.OrgID != t.OrgID.String() {
+		log.Warn().Str("user_org", claims.OrgID).Str("topic_org", t.OrgID.String()).Msg("user does not own topic")
 		c.JSON(http.StatusNotFound, api.ErrorResponse("topic not found"))
 		return
 	}
 
+	// Only archive mode is available on update
+	if topic.State == pb.TopicTombstone_READONLY.String() {
+		switch t.State {
+		case pb.TopicTombstone_READONLY:
+			c.JSON(http.StatusBadRequest, api.ErrorResponse("topic is already readonly"))
+			return
+		case pb.TopicTombstone_DELETING:
+			c.JSON(http.StatusBadRequest, api.ErrorResponse("topic is already being deleted"))
+			return
+		}
+
+		// Request one-time claims for the topic update request
+		req := &qd.Project{
+			ProjectID: t.ProjectID,
+		}
+		var rep *qd.LoginReply
+		if rep, err = s.quarterdeck.ProjectAccess(ctx, req); err != nil {
+			log.Error().Err(err).Msg("could not request one-time claims")
+			c.JSON(qd.ErrorStatus(err), api.ErrorResponse("could not update topic"))
+			return
+		}
+
+		// Create the Ensign context with the one-time claims
+		ensignContext := qd.ContextWithToken(ctx, rep.AccessToken)
+
+		// Archive means a "soft" delete in Ensign (no data is destroyed)
+		updateRequest := &pb.TopicMod{
+			Id:        t.ID.String(),
+			Operation: pb.TopicMod_ARCHIVE,
+		}
+		var tombstone *pb.TopicTombstone
+		if tombstone, err = s.ensign.DeleteTopic(ensignContext, updateRequest); err != nil {
+			if status.Code(err) == codes.NotFound {
+				log.Warn().Err(err).Str("topicID", updateRequest.Id).Msg("topic not found in ensign")
```

Comment:
```suggestion
				log.Error().Err(err).Str("topicID", updateRequest.Id).Msg("topic not found in ensign even though it is tenant")
```

Reply:
This seems like a serious condition right?

Reply:
Right, although potentially it could just be a syncing issue where the topic was just created in Tenant but it hasn't made its way through Ensign yet.